core/src/main/java/io/grpc/internal/SpiffeUtil.java

```diff
@@ -96,6 +123,140 @@ private static void validatePathSegment(String pathSegment) {
             + " ([a-zA-Z0-9.-_])");
   }
 
+  /**
+   * Returns the SPIFFE ID from the leaf certificate, if present.
+   *
+   * @param certChain certificate chain to extract SPIFFE ID from
+   */
+  public static Optional<SpiffeId> extractSpiffeId(X509Certificate[] certChain)
+      throws CertificateParsingException {
+    checkArgument(checkNotNull(certChain, "certChain").length > 0, "certChain can't be empty");
+    Collection<List<?>> subjectAltNames = certChain[0].getSubjectAlternativeNames();
+    if (subjectAltNames == null) {
+      return Optional.absent();
+    }
+    String uri = null;
+    // Search for the unique URI SAN.
+    for (List<?> altName : subjectAltNames) {
+      if (altName.size() < 2 ) {
+        continue;
+      }
+      if (URI_SAN_TYPE.equals(altName.get(0))) {
+        if (uri != null) {
+          throw new IllegalArgumentException("Multiple URI SAN values found in the leaf cert.");
+        }
+        uri = (String) altName.get(1);
+      }
+    }
+    if (uri == null) {
+      return Optional.absent();
+    }
+    return Optional.of(parse(uri));
+  }
+
+  /**
+   * Loads a SPIFFE trust bundle from a file, parsing it from the JSON format.
+   * In case of success, returns {@link SpiffeBundle}.
+   * If any element of the JSON content is invalid or unsupported, an
+   * {@link IllegalArgumentException} is thrown and the entire Bundle is considered invalid.
+   *
+   * @param trustBundleFile the file path to the JSON file containing the trust bundle
+   * @see <a href="https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Trust_Domain_and_Bundle.md">JSON format</a>
+   * @see <a href="https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#61-publishing-spiffe-bundle-elements">JWK entry format</a>
+   * @see <a href="https://datatracker.ietf.org/doc/html/rfc7517#appendix-B">x5c (certificate) parameter</a>
+   */
+  public static SpiffeBundle loadTrustBundleFromFile(String trustBundleFile) throws IOException {
+    Map<String, ?> trustDomainsNode = readTrustDomainsFromFile(trustBundleFile);
+    Map<String, List<X509Certificate>> trustBundleMap = new HashMap<>();
+    Map<String, Long> sequenceNumbers = new HashMap<>();
+    for (String trustDomainName : trustDomainsNode.keySet()) {
+      Map<String, ?> domainNode = JsonUtil.getObject(trustDomainsNode, trustDomainName);
+      if (domainNode == null || domainNode.size() == 0) {
```

Comment:
`null` is impossible, because that means the entry doesn't exist.

Reply:
Ditto for `kty` and `use`.

Reply:
I don't think I got it. For domainNode,
``` java
for (String trustDomainName : trustDomainsNode.keySet()) {
      Map<String, ?> domainNode = JsonUtil.getObject(trustDomainsNode, trustDomainName);
      if (domainNode == null || domainNode.size() == 0)
```
we iterate over the keys, and that's why `getObject` can't return null
For `kty` and `use`, we check if 

- jwkNode contains them
- the value matches the spec

so it is possible that for entries like 
```json
"keys": [
        {
          "x5c": "bytes"
        }
      ]
```
the values are null.
Am I missing something?

Reply:
I was thinking about how `getString()` does instanceof before returning. But yeah, if it the value is missing it will be null.

Reply:
I also realized that I didn't have any tests covering ClassCastException (there are few LOCs where it might happen). After adding one for  
```java
Long sequenceNumber = JsonUtil.getNumberAsLong(domainNode, "sequence_number");
```
I realized that I'm just piggybacking on `JsonUtil` exception pattern itself.
Yes, `JsonUtil` checks for key presence (and returns null if not found) before instanceof.